### PR TITLE
More Details in module level docstrings

### LINKED ISSUE #449: Add more details to module level docstrings
Due to the changes introduced in #448 and #440, the docstrings at the top of a `.py` file will appear in the documentation. Most of these docstrings lack details. 

For example,

image

https://github.com/vprusso/toqito/blob/48fbde1a462482e5c439c80e9e1234c6aa97c456/toqito/states/werner.py#L1

This also affects the docstring in `init` files. 

image

https://github.com/vprusso/toqito/blob/48fbde1a462482e5c439c80e9e1234c6aa97c456/toqito/states/__init__.py#L1

This issue is to add more details to each of the following modules:

- [ ] Add details to the module level docstrings: `toqito/channel_metrics`
- [ ] Add details to the module level docstrings: `toqito/channel_ops`
- [ ] Add details to the module level docstrings: `toqito/channel_props`
- [ ] Add details to the module level docstrings: `toqito/channels`
- [ ] Add details to the module level docstrings: `toqito/helper`
- [ ] Add details to the module level docstrings: `toqito/matrices`
- [ ] Add details to the module level docstrings: `toqito/matrix_ops`
- [ ] Add details to the module level docstrings: `toqito/matrix_props`
- [ ] Add details to the module level docstrings: `toqito/measurement_ops`
- [ ] Add details to the module level docstrings: `toqito/measurement_props`
- [ ] Add details to the module level docstrings: `toqito/nonlocal_games`
- [ ] Add details to the module level docstrings: `toqito/perms`
- [ ] Add details to the module level docstrings: `toqito/rand`
- [ ] Add details to the module level docstrings: `toqito/state_metrics`
- [ ] Add details to the module level docstrings: `toqito/state_ops`
- [ ] Add details to the module level docstrings: `toqito/state_opt`
- [ ] Add details to the module level docstrings: `toqito/state_props`
- [ ] Add details to the module level docstrings: `toqito/states`

### PR DESCRIPTION
**Pull request by Shivansh20128**
## Description
Provide a brief description of the PR's purpose here. If your PR is supposed to fix an existing issue, use
a [keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) to link your PR to the issue.
Resolves #449 

## Changes
Notable changes that this PR has either accomplished or will accomplish. Feel free to add more lines to the itemized list
below.

  -  [ ] Change 1
Adding docstrings to modules in toqito as mentioned in the issue.

## Checklist
Before marking your PR ready for review, make sure you checked the following locally. If this is your first PR, you might be notified of some workflow failures after a maintainer has approved the workflow jobs to be run on your PR. 

Additional information is available in the [documentation](https://toqito.readthedocs.io/en/latest/contributing.html#testing).

  -  [ ] Use `ruff` for errors related to code style and formatting.
  -  [ ] Verify all previous and newly added unit tests pass in `pytest`.
  -  [ ] Check the documentation build does not lead to any failures. `Sphinx` build can be checked locally for any failures related to your PR
  -  [ ] Use `linkcheck` to check for broken links in the documentation
  -  [ ] Use `doctest` to verify the examples in the function docstrings work as expected.

### CODE REVIEW

**Comment by Shivansh20128** [comment]
Hi! Can someone please look at the one docstring I have added for _Quantum States_, and give feedback? I will then accordingly add the docstrings for all the other modules. Please don't mind the ruff checks, for now, I will look at them later.
Thank You

File: toqito/states/__init__.py [hunk 1]
```diff
@@ -1,4 +1,7 @@
-"""This module refers to the `Quantum States` which acts as the parent to the below submodules. Quantum states are one of the three fundamental objects that `toqito` provides, the others being `channels` and `measurements`."""
+"""Quantum States is used to denote the module `Quantum State`. It acts as the parent to the below submodules.
```
Comment (vprusso) [inline toqito/states/__init__.py:1]: This isn't entirely true; there is no module called "Quantum State." Instead, this is a set of modules that numerically implement well-known quantum states in the literature.

Comment (Shivansh20128) [reply toqito/states/__init__.py:1]: Ohh okay. I will correct it.
Thank you

**Comment by Shivansh20128** [comment]
I have added and updated the docstrings for all modules and submodules. Please take a look at them.
Thank you

**Comment by Shivansh20128** [comment]
I think there is also some problem with the test for _tutorials.nonlocal_games.rst_. If you see the logs for this check above, you will see:
Screenshot 2024-10-11 at 1 15 43 PM

This check fails sometimes, and sometimes it doesn't, despite me not making any changes to this file.

**Comment by purva-thakre** [comment]
Moving the corresponding issue to the next milestone. Given the deadline, I won't be able to review this PR in time.

**Comment by purva-thakre** [comment]
> I think there is also some problem with the test for _tutorials.nonlocal_games.rst_. If you see the logs for this check above, you will see: Screenshot 2024-10-11 at 1 15 43 PM
> 
> This check fails sometimes, and sometimes it doesn't, despite me not making any changes to this file.

Yes, this is related to #613

**Comment by vprusso** [comment]
Hi @Shivansh20128, 

Yes, and I believe I know why this is an issue. 

Without going into too much detail, the lower bound technique can sometimes get stuck in a local optimum. The optimal value may indeed be lower than the actual optimal value (which is what we are seeing in this case here). 

Now, we could address this in a few ways. One way that might be the most sensible is to simply run the `np.around(chsh.quantum_value_lower_bound(),  decimals=2)` function multiple times (say 3-5 times) and then just take the maximum value of those runs. It will be highly unlikely (but not impossible) for the value to be lower than the true optimum value (0.85...). 

Another fix here would to simply not print out the value and give a brief explanation of what I said above. 

As @purva-thakre mentioned, this is captured in #613 so spinning up a separate issue with that fix should solve the problem.

Does that make sense? Let me know if you'd like to work on that issue as well, and thank you again for your work!

**Comment by Shivansh20128** [comment]
I am confused. When played using the concept of entanglement, I know that the chsh game gives the best probability of 0.85. The function `np.around(chsh.quantum_value_lower_bound(),  decimals=2)` is trying to calculate the lowest value of the quantum winning probability. If these two sentences are correct, then are we trying to _maximize the lower bound_?

**Comment by vprusso** [comment]
> I am confused. When played using the concept of entanglement, I know that the chsh game gives the best probability of 0.85. The function `np.around(chsh.quantum_value_lower_bound(), decimals=2)` is trying to calculate the lowest value of the quantum winning probability. If these two sentences are correct, then are we trying to _maximize the lower bound_?

Right, so let me unpack this a bit without getting too technical. 

Directly computing the optimal quantum value of a nonlocal game is not possible (as far as we know). Therefore, we have a set of heuristics to provide upper and lower bounds on this value. For the upper bound, we make use of the so-called NPA hierarchy (a hierarchy of SDPs that eventually converges on the optimal quantum value but scales exponentially and is NP-hard). This is numerically implemented [here](https://github.com/vprusso/toqito/blob/7b15429e1602dca93fdaf7d32f09f7ff697178a5/toqito/nonlocal_games/nonlocal_game.py#L543) via the `commuting_measurement_value_upper_bound` function. 

The lower bound (which is obtained through a "see-saw" technique of alternating projections between two SDPs). This is implemented numerically [here](https://github.com/vprusso/toqito/blob/7b15429e1602dca93fdaf7d32f09f7ff697178a5/toqito/nonlocal_games/nonlocal_game.py#L165) via the `quantum_value_lower_bound` function. This lower bound technique is not guaranteed to hit the optimal value. It can get stuck in a local optimum, and specifically, it can get stuck in a value that is lower than the optimum (which is what is happening here). 

Does that make sense?

**Comment by Shivansh20128** [comment]
Yes, now it does. And I just tried your proposed solution about running the function multiple times. I ran the code multiple times on my PC, and it gave 0.75 on the 22nd run.
So I think we can go ahead with this approach, and at the same time add a docstring there in case a user encounters the 0.75 value, explaining about about condition of being stuck in a local optimum.

image

**Comment by Shivansh20128** [comment]
I would enjoy working on this if you think it's okay.
And thank you so much for the deep insights and explanations. 
Thank you

**Comment by vprusso** [comment]
> and it gave 0.75 on the 22nd run.

Hopefully it gave ~0.85 on the 2nd run (or at least, I *hope* that's what you meant to say :) )

> I would enjoy working on this if you think it's okay.

Of course! Your contributions are sincerely and greatly appreciated! Thank you, @Shivansh20128 !

**Comment by Shivansh20128** [comment]
> Hopefully it gave ~0.85 on the 2nd run (or at least, I _hope_ that's what you meant to say :) )
Yes, it gave 0.85 till the 21st run.

**Comment by vprusso** [comment]
Looks like there may be some conflicts, but I think once those are resolved, this should be good to go! @Shivansh20128

Thank you again for your contributions to the project, this is great work!

**Comment by Shivansh20128** [comment]
@vprusso Done!

**Comment by vprusso** [comment]
Excellent work, @Shivansh20128 !

**Pull request merged**